File: src/subside/plugins/koth/scheduler/Schedule.java
package subside.plugins.koth.scheduler;

import lombok.Getter;
import lombok.Setter;
import org.json.simple.JSONObject;
import subside.plugins.koth.exceptions.KothException;
import subside.plugins.koth.modules.ConfigHandler;
import subside.plugins.koth.utils.JSONSerializable;
import subside.plugins.koth.utils.MessageBuilder;

import java.util.List;
import java.util.logging.Level;

public class Schedule implements JSONSerializable<Schedule> {
    private long nextEventMillis;
    private @Getter @Setter String koth;
    private @Getter @Setter int captureTime = 15;
    private @Getter @Setter Day day;
    private @Getter @Setter String time;
    private @Getter @Setter int maxRunTime = -1;
    private @Getter @Setter int lootAmount = -1;
    private @Getter @Setter String lootChest;
    private @Getter @Setter String entityType;

    private int preBroadcastIndex = 0;
    
    private static final long WEEK = 7 * 24 * 60 * 60 * 1000;
    
    private @Getter ScheduleHandler scheduleHandler;

    public Schedule(ScheduleHandler scheduleHandler){
        this.scheduleHandler = scheduleHandler;
    }
    
    public Schedule(ScheduleHandler scheduleHandler, String koth, Day day, String time) {
        this.scheduleHandler = scheduleHandler;
        this.koth = koth;
        this.day = day;
        this.time = time;
        calculateNextEvent();
    }

    /**
     * Calculate when the next event should happen
     */
    public void calculateNextEvent() {
        long eventTime = day.getDayStart(scheduleHandler.getPlugin()) + Day.getTime(time) - WEEK;
        
        eventTime += scheduleHandler.getPlugin().getConfigHandler().getGlobal().getScheduleMinuteOffset()*60*1000;

        while (eventTime < System.currentTimeMillis()) {
            eventTime += WEEK;
        }
        nextEventMillis = eventTime;
        
        if(scheduleHandler.getPlugin().getConfigHandler().getGlobal().isDebug()){
            scheduleHandler.getPlugin().getLogger().log(Level.WARNING, "Schedule created for: "+day+" "+time+" "+nextEventMillis);
        }

        // Broadcast stuff
        List<Integer> times = scheduleHandler.getPlugin().getConfigHandler().getGlobal().getPreBroadcastTimes();
        this.preBroadcastIndex = times.size()-1;
        while(preBroadcastIndex >= 0 && (nextEventMillis - times.get(preBroadcastIndex) * 1000 < System.currentTimeMillis())){
            preBroadcastIndex--;
        }
    }

    /**
     * This method is executed every second,
     * it checks if it should broadcast anything and if it should run
     */
    public void tick() {
        // Broadcast stuff
        ConfigHandler.Global cGlobal = scheduleHandler.getPlugin().getConfigHandler().getGlobal();
        if (cGlobal.isPreBroadcast() && preBroadcastIndex >= 0) {
            int time = cGlobal.getPreBroadcastTimes().get(preBroadcastIndex);
            long nextBroadcast = nextEventMillis - time * 1000;
            if (System.currentTimeMillis() > nextBroadcast) {
                new MessageBuilder(cGlobal.getPreBroadcastMessages().get(time)).maxTime(maxRunTime).captureTime(captureTime).lootAmount(lootAmount).koth(scheduleHandler.getPlugin().getKothHandler(), koth).buildAndBroadcast();
                preBroadcastIndex--;
            }
        }

        // Code for when a KoTH should start
        if (System.currentTimeMillis() > nextEventMillis) {
            // Reset prebroadcasting
            preBroadcastIndex = scheduleHandler.getPlugin().getConfigHandler().getGlobal().getPreBroadcastTimes().size()-1;

            // Set the next event time
            setNextEventTime();
            try {
                scheduleHandler.getPlugin().getKothHandler().startKoth(this);
            } catch(KothException e){
                scheduleHandler.getPlugin().getLogger().warning("Koth is already running");
            }
        }
    }

    private void setNextEventTime() {
        nextEventMillis += WEEK;
    }

    /**
     * Simple getter for nextEventMillis
     * @return When the next event should run (in milliseconds)
     */
    public long getNextEvent() {
        return nextEventMillis;
    }

    public Schedule load(JSONObject obj) {
        this.koth =  (String) obj.get("koth"); // koth
        this.day = Day.getDay((String)obj.get("day")); // day
        this.time = (String) obj.get("time"); // time
        
        
        if (obj.containsKey("captureTime")) {
            this.setCaptureTime(Integer.parseInt(obj.get("captureTime")+"")); // runTime
        }

        if (obj.containsKey("maxRunTime")) {
            this.setMaxRunTime(Integer.parseInt(obj.get("maxRunTime")+"")); // maxRunTime
        }

        if (obj.containsKey("lootAmount")) {
            this.setLootAmount(Integer.parseInt(obj.get("lootAmount")+"")); // lootAmount
        }

        if (obj.containsKey("lootChest")) {
            this.setLootChest((String) obj.get("lootChest")); // lootChest
        }
        
        if(obj.containsKey("entityType")){
            this.setEntityType((String) obj.get("entityType"));
        }

        calculateNextEvent();
        return this;

    }

    @SuppressWarnings("unchecked")
    public JSONObject save() {
        JSONObject obj = new JSONObject();
        obj.put("koth", this.koth); // koth
        obj.put("day", this.day.getDay()); // day
        obj.put("time", this.time); // time

        if (captureTime != -1) {
            obj.put("captureTime", this.captureTime); // runTime
        }

        if (maxRunTime != -1) {
            obj.put("maxRunTime", this.maxRunTime); // maxRunTime
        }

        if (lootAmount != -1 || lootAmount == scheduleHandler.getPlugin().getConfigHandler().getLoot().getLootAmount()) {
            obj.put("lootAmount", this.lootAmount); // lootAmount
        }

        if (lootChest != null) {
            obj.put("lootChest", this.lootChest); // lootChest
        }
        
        if(entityType != null){
            obj.put("entityType", this.entityType);
        }

        return obj;
    }
}

File: src/subside/plugins/koth/scheduler/Schedule.kt
package subside.plugins.koth.scheduler

import org.bukkit.configuration.serialization.ConfigurationSerializable
import subside.plugins.koth.KothManager
import subside.plugins.koth.utils.MessageBuilder
import java.util.logging.Level

class Schedule(
    var koth: String,
    var day: Day,
    var time: String
) : ConfigurationSerializable {
    var captureTime = 15
    var maxRunTime = -1
    var lootAmount = -1
    var lootChest: String? = null
    var entityType: String? = null

    private var nextEventMillis: Long = 0
    private var preBroadcastIndex: Int = 0
    private var hasSetup = false

    fun tick(kothManager: KothManager) {
        if (!hasSetup) {
            calculateNextEvent(kothManager)
            hasSetup = true
        }

        // Broadcast stuff
        val configGlobal = kothManager.configHandler.global
        // If we have anything to broadcast
        if (configGlobal.preBroadcast && preBroadcastIndex >= 0) {
            // We calculate the next broadcast time
            val time = configGlobal.preBroadcastTimes[preBroadcastIndex]
            val nextBroadcast = nextEventMillis - time * 1000
            // If we the next broadcast time already past we should broadcast it
            if (System.currentTimeMillis() > nextBroadcast) {
                MessageBuilder(configGlobal.preBroadcastMessages[time])
                    .maxTime(maxRunTime)
                    .captureTime(captureTime)
                    .lootAmount(lootAmount)
                    .koth(kothManager.kothHandler, koth)
                    .broadcast()
                preBroadcastIndex--
            }
        }

        // Code for when a KoTH should start
        if (System.currentTimeMillis() > nextEventMillis) {
            // We reset the broadcast index
            preBroadcastIndex = kothManager.configHandler.global.preBroadcastTimes.size - 1

            // And we set the next event time
            setNextEventTime()
            try {
                kothManager.kothHandler.startKoth(this)
            } catch (e: Exception) {
                kothManager.plugin.get()!!.logger.warning("KoTH is already running")
            }
        }
    }

    private fun setNextEventTime() {
        nextEventMillis += WEEK
    }

    /**
     * Simple getter for nextEventMillis
     * @return When the next event should run (in milliseconds)
     */
    fun getNextEvent(): Long {
        return nextEventMillis
    }


    private fun calculateNextEvent(kothManager: KothManager) {
        var eventTime = day.getDayStart(kothManager) + Day.getTime(time) - WEEK
        eventTime += kothManager.configHandler.global.scheduleMinuteOffset * 60 * 1000

        while (eventTime < System.currentTimeMillis()) {
            eventTime += WEEK
        }

        if (kothManager.configHandler.global.debug) {
            kothManager.plugin.get()!!.logger.log(Level.WARNING, "Schedule created for: $day $time $nextEventMillis")
        }

        // Broadcast stuff
        val times = kothManager.configHandler.global.preBroadcastTimes
        preBroadcastIndex = times.size - 1
        while (preBroadcastIndex >= 0 && nextEventMillis - times[preBroadcastIndex] * 1000 < System.currentTimeMillis()) {
            preBroadcastIndex--
        }
    }

    override fun serialize(): Map<String, Any?> {
        return mapOf(
            "koth" to koth,
            "day" to day.identifier,
            "time" to time,
            "captureTime" to captureTime,
            "maxRunTime" to maxRunTime,
            "lootAmount" to lootAmount,
            "lootChest" to lootChest,
            "entityType" to entityType
        )
    }

    companion object {
        const val WEEK: Long = 7 * 24 * 60 * 60 * 1000

        @JvmStatic fun deserialize(args: Map<String, Any>): Schedule {
            return Schedule(
                args["koth"] as String,
                Day.values().find { it.identifier == args["day"] } ?: Day.MONDAY,
                args["time"] as String
            ).also {
                it.captureTime = args.getOrDefault("captureTime", 15) as Int
                it.maxRunTime = args.getOrDefault("maxRunTime", -1) as Int
                it.lootAmount = args.getOrDefault("lootAmount", -1) as Int
                it.lootChest = args.getOrDefault("lootChest", null) as String?
                it.entityType = args.getOrDefault("entityType", null) as String?
            }
        }
    }
}